packages/typir/src/utils/utils-definitions.ts

```diff
@@ -29,44 +37,479 @@ export function isNameTypePair(type: unknown): type is NameTypePair {
+     * In order to solve these cycles, types which are part of such "dependency cycles" should be ignored during waiting,
+     * e.g. A should not waiting B and B should not wait for A.
+     * These types to ignore are stored in the following Set.
+     */
+    protected typesToIgnoreForCycles: Set<Type> = new Set();
+
+    /** These TypeReferences must be in the states Identifiable or Completed, before the listeners are informed */
+    protected readonly waitForRefsIdentified: Array<TypeReference<T>> | undefined;
+    /** These TypeReferences must be in the state Completed, before the listeners are informed */
+    protected readonly waitForRefsCompleted: Array<TypeReference<T>> | undefined;
+
+    /** These listeners will be informed once, when all TypeReferences are in the desired state.
+     * If some of these TypeReferences are invalid (the listeners will not be informed about this) and later in the desired state again,
+     * the listeners will be informed again, and so on. */
+    protected readonly listeners: Array<WaitingForIdentifiableAndCompletedTypeReferencesListener<T>> = [];
+
+    constructor(
+        waitForRefsToBeIdentified: Array<TypeReference<T>> | undefined,
+        waitForRefsToBeCompleted: Array<TypeReference<T>> | undefined,
+    ) {
+
+        // remember the relevant TypeReferences to wait for
+        this.waitForRefsIdentified = waitForRefsToBeIdentified;
+        this.waitForRefsCompleted = waitForRefsToBeCompleted;
+
+        // register to get updates for the relevant TypeReferences
+        toArray(this.waitForRefsIdentified).forEach(ref => ref.addListener(this, true)); // 'true' calls 'checkIfFulfilled()' to check, whether everything might already be fulfilled
+        toArray(this.waitForRefsCompleted).forEach(ref => ref.addListener(this, true));
+    }
+
+    deconstruct(): void {
+        this.listeners.splice(0, this.listeners.length);
+        this.waitForRefsIdentified?.forEach(ref => ref.removeListener(this));
+        this.waitForRefsCompleted?.forEach(ref => ref.removeListener(this));
+        this.typesToIgnoreForCycles.clear();
+    }
+
+    addListener(newListener: WaitingForIdentifiableAndCompletedTypeReferencesListener<T>, informAboutCurrentState: boolean): void {
+        this.listeners.push(newListener);
+        // inform the new listener
+        if (informAboutCurrentState) {
+            if (this.fulfilled) {
+                newListener.onFulfilled(this);
+            } else {
+                newListener.onInvalidated(this);
+            }
+        }
+    }
+
+    removeListener(listenerToRemove: WaitingForIdentifiableAndCompletedTypeReferencesListener<T>): void {
+        const index = this.listeners.indexOf(listenerToRemove);
+        if (index >= 0) {
+            this.listeners.splice(index, 1);
+        }
+    }
+
+    addTypesToIgnoreForCycles(moreTypesToIgnore: Set<Type>): void {
```

Comment:
This one could use a little more commenting on what it is good for, I think.

Reply:
done